staging/src/k8s.io/client-go/util/workqueue/queue.go

```diff
@@ -252,53 +256,16 @@ func (q *Type) ShutDown() {
 // ShutDownWithDrain, as to force the queue shut down to terminate immediately
 // without waiting for the drainage.
 func (q *Type) ShutDownWithDrain() {
-	q.setDrain(true)
-	q.shutdown()
-	for q.isProcessing() && q.shouldDrain() {
-		q.waitForProcessing()
-	}
-}
-
-// isProcessing indicates if there are still items on the work queue being
-// processed. It's used to drain the work queue on an eventual shutdown.
-func (q *Type) isProcessing() bool {
-	q.cond.L.Lock()
-	defer q.cond.L.Unlock()
-	return q.processing.len() != 0
-}
-
-// waitForProcessing waits for the worker goroutines to finish processing items
-// and call Done on them.
-func (q *Type) waitForProcessing() {
-	q.cond.L.Lock()
-	defer q.cond.L.Unlock()
-	// Ensure that we do not wait on a queue which is already empty, as that
-	// could result in waiting for Done to be called on items in an empty queue
-	// which has already been shut down, which will result in waiting
-	// indefinitely.
-	if q.processing.len() == 0 {
-		return
-	}
-	q.cond.Wait()
-}
-
-func (q *Type) setDrain(shouldDrain bool) {
-	q.cond.L.Lock()
-	defer q.cond.L.Unlock()
-	q.drain = shouldDrain
-}
-
-func (q *Type) shouldDrain() bool {
 	q.cond.L.Lock()
 	defer q.cond.L.Unlock()
-	return q.drain
-}
 
-func (q *Type) shutdown() {
-	q.cond.L.Lock()
-	defer q.cond.L.Unlock()
+	q.drain = true
```

Comment:
What happens if someone calls ‘Shutdown’ followed by ‘ShutdownWithDrain’? Should this function check if the queue is already shutting down?

Reply:
I didn't want to implement the one line fix because this would then be checking the condition, calling a method, and then checking the same condition again so I think it is worth making a small code improvement with the fix (and making this tighter reduces the scope for another race). 

> What happens if someone calls ‘Shutdown’ followed by ‘ShutdownWithDrain’? 
Hmm that's an interesting one, but as 'Shutdown' is pretty much immediate, so it won't enter any loop.  Once someone calls it, we expect all waiting clients to wake up and go-away and then presumably the return back to the caller (presumably to continue exiting).

Reply:
> Hmm that's an interesting one, but as 'Shutdown' is pretty much immediate. 

I guess I’m asking if it is possible for a race to occur by Shutdown being called and not being respected due to false ‘drain’ not being observed by the waking goroutines before ShutdownWithDrain is called. I would expect the first call to take precedence.

> This is in keeping with the existing code behaviour.

yep, just curious if this is another potential bug

Reply:
Both 'Shutdown' and 'ShutDownWithDrain' have the identical effect on the workers - which are told to wake up with an explicit expectation that they also check the shuttingDown bool and act accordingly.  Any new items are rejected by the queue (because it is shutting down).

The only difference between the two is that 'ShutDownWithDrain' will hang around until all of the existing items have been consumed off the queue.  This fix does not change this behaviour - all it does it move the locking around a bit so that 'ShutDown' and 'ShutDownWithDrain` are treated a little bit more like a single transaction - as all of the other methods are implemented - i.e. getting the lock before they perform an atomic action.

Reply:
Gotcha I see now drain is not exposed and ShutdownWithDrain only adds waiting. Can mark this comment resolved